js/spreadsheet.js

```diff
@@ -67,248 +67,109 @@ var spreadsheet = function(options) {
     }
   }
 
-  /**
-   * We user this method to determine where is closest data is lokated from the selected cell
-   * 
-   * @param {Array} selectedCell - array of the coordinat of the selected cell reacived throw the hot.getSelected() method
-   * 
-   * @returns {Object} - object of the directions where data is placed occording to the selected cell
-   */
-  function closestData(selectedCell) {
-    selectedCell = selectedCell[0];
-    if (!Array.isArray(selectedCell)) {
-      console.error('We must pass an array of the cell coordinats to the closestData function. First element is cell' +
-        'row and second element is cell col. In this case script will act as if there was a value in the cell. ' +
-        'Value that was passed - ',
-        selectedCell);
-      return false;
-    }
-
-    var col = selectedCell[1];
-    var row = selectedCell[0];
-    var selectedCellData = hot.getDataAtCell(row, col)
-    // At this block we getting an index of the nearest cells from the selected cell
-    // If we selected first row it ID is 0 already and if we - 1 from it we will reacive an error in the hot.getDataAtCell() method
-    var top = row ? row - 1 : row;
-    var bottom = row + 1;
-    // Same as in top variable
-    var left = col ? col - 1 : col;
-    var right = col + 1;
-    // At this block we reacive the nearest cells value from the selected cell
-    var leftValue = hot.getDataAtCell(row, left);
-    var rightValue = hot.getDataAtCell(row, right);
-    var topValue = hot.getDataAtCell(top, col);
-    var bottomValue = hot.getDataAtCell(bottom, col);
-    var dataAt = {
-      left: false,
-      right: false,
-      top: false,
-      bottom: false,
-      all: false,
-      hasData: false
-    };
-
-    // If there is a data in the selected cell we should select data releated to this cell
-    if (selectedCellData !== null) {
-      dataAt.hasData = true;
-      return dataAt;
-    }
-
-    // If no value near the selected cell we should select all table, also fires when we just load data source
-    // and clicked ctrl+a combination
-    if (leftValue === null && rightValue === null && topValue === null && bottomValue === null) {
-      dataAt.all = true;
-      return dataAt;
-    }
-
-    // Showing where is data position from the selected cell
-    dataAt.left = leftValue !== null;
-    dataAt.right = rightValue !== null;
-    dataAt.top = topValue !== null;
-    dataAt.bottom = bottomValue !== null;
+  function getNearestCells(cellPosition) {
+    var leftShift = cellPosition[1] - 1;
+    var topShift = cellPosition[0] - 1;
+    var left = [cellPosition[0], leftShift < 0 ? cellPosition[1] : leftShift];
+    var right = [cellPosition[0], cellPosition[1] + 1];
+    var top = [topShift < 0 ? cellPosition[0] : topShift, cellPosition[1]];
+    var bottom = [cellPosition[0] + 1, cellPosition[1]];
 
-    return dataAt;
+    return [left, right, top, bottom];
```

Comment:
Could we not be more specific like this in all this new code? The current format is more complex to read, understand and maintain.

```suggestion
    return {
      top: top,
      right: right,
      bottom: bottom,
      left: left
    };
```

Reply:
@tonytlwu We could do it in this function, yes, but not for this [function](https://github.com/Fliplet/fliplet-widget-data-source/pull/75/files#diff-58f9b9f8183cba2a53b4728f2df11588R127) because such format needed for the `handsontable` library.

Reply:
@tonytlwu also about this return we decided to use this format:
1. It is easy to expand functionality for this feature when we decide to look for data in diagonal.
2. This return used in a sequence of the functions we have written and we are expecting an array to be a result of this function. So if we change it to the object we will have to transform it back to the array after that

Reply:
1. What do you mean by deciding to look for data in diagonal? Can you give me code examples of what it would roughly look in one method compared to another?
2. I think if we use `.top` and other properties, it's essentially only in this line https://github.com/Fliplet/fliplet-widget-data-source/pull/75/files#diff-58f9b9f8183cba2a53b4728f2df11588R369 that it needs to be converted to an array format (because handsontable needs it). I don't like the fact that all these small functions are written in a format handsontable dictates when it can just be a helper function that transforms these in the moment right before handsontable needs them. (We don't necessarily have to change the format now, as I'd like to see what you mean by point 1 first to decide.)

Reply:
> What do you mean by deciding to look for data in diagonal? Can you give me code examples of what it would roughly look in one method compared to another?

By diagonal I mean that now we look for data on top, left, bottom and right to decide should we what range we should select. But if we will want to look for data at top-left, top-right bottom-left and bottom-right in the current function we will simply add these cells to the array but with the suggested function, we will need to add an extra object key and search for it in next functions.
For example changes with our function:
```
  function getNearestCells(cellPosition) {
    var leftShift = cellPosition[1] - 1;
    var topShift = cellPosition[0] - 1;
    var left = [cellPosition[0], leftShift < 0 ? cellPosition[1] : leftShift];
    var right = [cellPosition[0], cellPosition[1] + 1];
    var top = [topShift < 0 ? cellPosition[0] : topShift, cellPosition[1]];
    var bottom = [cellPosition[0] + 1, cellPosition[1]];
    var topLeft = [topShift < 0 ? cellPosition[0] : topShift, leftShift < 0 ? cellPosition[1] : leftShift];
    var topRight = [cellPosition[0] + 1, cellPosition[1]];

    return [left, right, top, bottom, topLeft, topRight ];
  }
```

Changes with an object for top value only.

```
  function getNearestCells(cellPosition) {
    var leftShift = cellPosition[1] - 1;
    var topShift = cellPosition[0] - 1;
    var left = [cellPosition[0], leftShift < 0 ? cellPosition[1] : leftShift];
    var right = [cellPosition[0], cellPosition[1] + 1];
    var top = [topShift < 0 ? cellPosition[0] : topShift, cellPosition[1]];
    var bottom = [cellPosition[0] + 1, cellPosition[1]];
    var topLeft = [topShift < 0 ? cellPosition[0] : topShift, leftShift < 0 ? cellPosition[1] : leftShift];
    var topRight = [cellPosition[0] + 1, cellPosition[1]];

    return {
      left: left,
      right: right,
      top: top,
      bottom: bottom,
      topLeft: topLeft,
      topRight: topRight
    };
  }

  function getDataCoordinates(tableData, cellPosition, processedCells) {
    if (!processedCells) {
      processedCells = [];
      processedCells[cellPosition[0]] = [];
      processedCells[cellPosition[0]][cellPosition[1]] = true;
    }

    var cellsGroup = getNearestCells(cellPosition);

    cellsGroup['currentCell'] = cellPosition;

    var processedRow;
    var processedCell;

    if (tableData[cellsGroup.top] !== null) {
      processedRow = processedCells[cellsGroup.top[0]];

      if (!processedRow) {
        processedCells[cellsGroup.top[0]] = [];
      }

      processedCell = processedCells[cellsGroup.top[0]][cellsGroup.top[1]];

      if (!processedCell) {
        processedCells[cellsGroup[cell][0]][cellsGroup[cell][1]] = false;
      }
      
      getDataCoordinates(tableData, cellsGroup.top, )
    }

    // And for every key in the object, we will need such if
}
```

Example with `for ... in` : 
```
  function getDataCoordinates(tableData, cellPosition, processedCells) {
    if (!processedCells) {
      processedCells = [];
      processedCells[cellPosition[0]] = [];
      processedCells[cellPosition[0]][cellPosition[1]] = true;
    }

    var cellsGroup = getNearestCells(cellPosition);

    cellsGroup['currentCell'] = cellPosition;

    for (var cell in cellsGroup) {
      var hasData = tableData[cellsGroup[cell][0]][cellsGroup[cell][1]] !== null;
      var processedRow = processedCells[cellsGroup[cell][0]];

      if (!processedRow) {
        processedCells[cellsGroup[cell][0]] = [];
      }

      var processedCell = processedCells[cellsGroup[cell][0]][cellsGroup[cell][1]];

      if (!processedCell) {
        processedCells[cellsGroup[cell][0]][cellsGroup[cell][1]] = false;
      }

      if (hasData && (!processedRow || !processedCell)) {
        processedCells[cellsGroup[cell][0]][cellsGroup[cell][1]] = true;
        getDataCoordinates(tableData, cell, processedCells);
      }      
    }
```

Reply:
@YaroslavOvdii I think I'm understanding this a little better now. Much of the confusion is created by the `[0]`, `[1]` notations and ambiguous naming, which made it difficult to understand.

Am I right in assuming that `getNearestCells()` takes the position of one cell and returns a range of cells, e.g. from A1 to E5? If so, then:

1. I still recommend using objects and properties instead of an array of numbers, which is unclear.
1. The range would be defined by either:
    1. `{ topLeft, topRight, bottomRight, bottomLeft }` or
    1. `{ top, left, columns, rows }` **Preferred**

Reply:
@tonytlwu 
> Am I right in assuming that getNearestCells() takes the position of one cell and returns a range of cells, e.g. from A1 to E5? If so, then:

No, it is not a range. It's simply coordinates of the four cells around one cell. And we won't be able to refuse from the `[0]` and `[1]` because the cell position is an array `[rowIndex, colIndex]` and we needed this coordinates to find the value of the cell in the `tableData` array.
I'll try to remade as

Reply:
@YaroslavOvdii I see. Then I think `{ top, left, bottom, right }` is the right way to go. If "top left" was needed, I'd just take the x coordinate of one and y coordinate of the other when it is needed, without needing to provide additional properties.

Reply:
@tonytlwu 
If you don't mind I would describe a full flow here.

The main idea was that we will serf threw the cells of the data source to determine a range of data we need to select.

Step 1.
Get selected cell coordinates.
![image](https://user-images.githubusercontent.com/53430352/78364615-6bc0e700-75c6-11ea-90e1-ca66b5ca6c0d.png)

Step 2.
Get the nearest cells top, bottom, left, right.
![image](https://user-images.githubusercontent.com/53430352/78364742-a75bb100-75c6-11ea-8dfa-428192448998.png)

Step 3.
Find if there is a data in the cells nearby to the selected cell.
If we have data in any cell from Step 2, we set it coordinates for Step 1 and repeat these steps until we won't get all cells with data scanned, but with a condition that it doesn't have to be a space between cells with data.
So in our case scan will go left, right, but won't go top and bottom because there is no data in those cells.

Step 4.
When we get a scanned cells array which looks like this `[rowIndex][cellIndex] = hasData true/false`  we simply get `max` and `min` values of the `rowsIndex` and `cellIndex`  which has `true` that is our range that we need to select for the user.

Reply:
@tonytlwu added changes as suggested.